### PR TITLE
nrf_802154: integrate sdk-nrf-802154 repository

### PR DESCRIPTION
**Pull request by e-rk**
To checkout:
```
pip install west==0.10.0a1
west config manifest.group-filter +nrf-802154
west update 
```

A kconfig choice selection is available: `NRF_802154_SOURCE`. To build with sdk-nrf-802154 select `internal`.
Other choices are `nrfxlib` and `Open source`.

### CODE REVIEW

File: modules/nrfxlib/nrf_802154/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      ${ZEPHYR_CURRENT_MODULE_DIR}/nrf_802154/sl/platform/hp_timer/nrf_802154_hp_timer.c
+      ${ZEPHYR_CURRENT_MODULE_DIR}/nrf_802154/sl/platform/lp_timer/nrf_802154_lp_timer.c
+  )
+endif ()
+
+if (CONFIG_MPSL)
+  target_compile_definitions(nrf-802154-platform PUBLIC NRF_802154_VERIFY_PERIPHS_ALLOC_AGAINST_MPSL=1)
+endif()
+
+add_subdirectory(${ZEPHYR_CURRENT_MODULE_DIR}/nrf_802154 nrf_802154)
```
Comment (tejlmand) [inline modules/nrfxlib/nrf_802154/CMakeLists.txt:21]: Please indent this section.

File: modules/nrfxlib/nrf_802154/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+if (CONFIG_NRF_802154_RADIO_DRIVER OR CONFIG_NRF_802154_SERIALIZATION)
+
+if (NOT CONFIG_NRF_802154_SL_OPENSOURCE AND NOT CONFIG_SOC_SERIES_NRF53X)
+  target_sources(nrf-802154-platform
```
Comment (tejlmand) [inline modules/nrfxlib/nrf_802154/CMakeLists.txt:10]: the `nrf-802154-platform`is defined in `zephyr` itself, and then sources are added here in `sdk-nrf` but include directories and link libraries are specified in `sdk-nrfxlib`.

Would it be better to have a variable to control use of the platform lib, similar to what is done for radio and serilization ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
Initial review.

File: modules/modules.cmake [hunk 3]
```diff
@@ -2,3 +2,4 @@ set(ZEPHYR_MCUBOOT_CMAKE_DIR ${CMAKE_CURRENT_LIST_DIR}/mcuboot)
 set(ZEPHYR_MCUBOOT_KCONFIG   ${CMAKE_CURRENT_LIST_DIR}/mcuboot/Kconfig)
 set(ZEPHYR_CDDL_GEN_CMAKE_DIR ${CMAKE_CURRENT_LIST_DIR}/cddl-gen)
 set(ZEPHYR_CDDL_GEN_KCONFIG   ${CMAKE_CURRENT_LIST_DIR}/cddl-gen/Kconfig)
+set(ZEPHYR_NRFXLIB_CMAKE_DIR ${CMAKE_CURRENT_LIST_DIR}/nrfxlib)
```
[the remark this answers is not in the document]
Comment (e-rk) [reply modules/modules.cmake]: Fixed

File: modules/nrfxlib/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+add_subdirectory(${ZEPHYR_CURRENT_MODULE_DIR} nrfxlib)
```
[the remark this answers is not in the document]
Comment (e-rk) [reply modules/nrfxlib/CMakeLists.txt:7]: The change that modified nrfxlib revision was merged with the commit that introduces this file.

File: modules/nrfxlib/nrf_802154/CMakeLists.txt (hunk 1 above)
Comment (e-rk) [reply modules/nrfxlib/nrf_802154/CMakeLists.txt]: Done

File: modules/nrfxlib/nrf_802154/CMakeLists.txt (hunk 2 above)
Comment (e-rk) [reply modules/nrfxlib/nrf_802154/CMakeLists.txt]: The problem with those source files (namely the `nrf_802154_hp_timer.c` and `nrf_802154_lp_timer.c`) is that those files are never used in vanilla zephyr, therefore I thought that it is better add them only in NCS glue. I have made sure that https://github.com/zephyrproject-rtos/zephyr/blob/51efef4ff24b04a523bb7da9a11b6484e1248133/modules/hal_nordic/CMakeLists.txt will always be evaluated before this file by adding the `depends` statement to https://github.com/nrfconnect/sdk-nrfxlib/blob/4b15172170e628d0059d754532f72117bf02a63a/zephyr/module.yml.

**Comment by e-rk** [comment]
I have addressed some of the comments and made a rebase.

File: modules/nrfxlib/nrf_802154/CMakeLists.txt (hunk 2 above)
Comment (tejlmand) [reply modules/nrfxlib/nrf_802154/CMakeLists.txt]: thanks for the explanation.

**Review by tejlmand — APPROVED** [review]

**Review by tejlmand — CHANGES_REQUESTED** [review]
During work on the CMakeLists code, I observed the following.
Building `samples/openthread/cli/` works:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild `samples/openthread/cli/
$ ninja -Cbuild
...
[538/538] Linking CXX executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      463472 B         1 MB     44.20%
            SRAM:       95168 B       256 KB     36.30%
        IDT_LIST:          0 GB         2 KB      0.00%
```

Switching to the opensource on cli also works:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild samples/openthread/cli/ -DCONFIG_NRF_802154_SL_OPENSOURCE=y -DCONFIG_NRF_802154_SOURCE_HAL_NORDIC=y
$ ninja -Cbuild
...
[536/536] Linking CXX executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      437692 B         1 MB     41.74%
            SRAM:       91840 B       256 KB     35.03%
        IDT_LIST:          0 GB         2 KB      0.00%
```

but if using:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild `samples/openthread/cli/
$ ninja -Cbuild menuconfig
# Set:
# NRF_802154_SL_OPENSOURCE=y
# NRF_802154_SOURCE_HAL_NORDIC=y
# results in
$ ninja -Cbuild
...
/projects/github/ncs/nrfxlib/mpsl/lib/cortex-m4/hard-float/libmpsl.a(libmpsl_debug_hardfp__obfuscated.elf): In function `nrf_802154_fal_tx_power_get':
:(.text+0x0): multiple definition of `nrf_802154_fal_tx_power_get'
modules/hal_nordic/nrf_802154/sl/sl_opensource/libnrf-802154-sl.a(nrf_802154_sl_fem.c.obj):/projects/github/ncs/modules/hal/nordic/drivers/nrf_802154/sl/sl_opensource/src/nrf_802154_sl_fem.c:128: first defined here
/projects/github/ncs/nrfxlib/mpsl/lib/cortex-m4/hard-float/libmpsl.a(libmpsl_debug_hardfp__obfuscated.elf): In function `mpsl_fem_abort_clear':
:(.text+0x0): multiple definition of `mpsl_fem_abort_clear'
modules/hal_nordic/nrf_802154/sl/sl_opensource/libnrf-802154-sl.a(nrf_802154_sl_fem.c.obj):nrf_802154_sl_fem.c:(.text.mpsl_fem_abort_clear+0x0): first defined here
/projects/github/ncs/nrfxlib/mpsl/lib/cortex-m4/hard-float/libmpsl.a(libmpsl_debug_hardfp__obfuscated.elf): In function `mpsl_fem_abort_extend':
:(.text+0x0): multiple definition of `mpsl_fem_abort_extend'
modules/hal_nordic/nrf_802154/sl/sl_opensource/libnrf-802154-sl.a(nrf_802154_sl_fem.c.obj):/projects/github/ncs/modules/hal/nordic/drivers/nrf_802154/sl/sl_opensource/src/nrf_802154_sl_fem.c:88: first defined here
/projects/github/ncs/nrfxlib/mpsl/lib/cortex-m4/hard-float/libmpsl.a(libmpsl_debug_hardfp__obfuscated.elf): In function `mpsl_fem_abort_reduce':
:(.text+0x0): multiple definition of `mpsl_fem_abort_reduce'
.....
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

so something is wrong when re-configuring using `menuconfig`.
Could this be related to some Kconfig dependencies / selects / values not being updated correctly ?

Also, selecting only `CONFIG_NRF_802154_SOURCE_HAL_NORDIC=y` also results in an error:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild samples/openthread/cli/ -DCONFIG_NRF_802154_SOURCE_HAL_NORDIC=y
...
CMake Error at /projects/github/ncs/zephyr/CMakeLists.txt:698 (add_dependencies):
  Cannot add target-level dependencies to non-existent target
  "nrf-802154-sl".

  The add_dependencies works for top-level logical targets created by the
  add_executable, add_library, or add_custom_target commands.  If you want to
  add file-level dependencies see the DEPENDS option of the add_custom_target
  and add_custom_command commands.
```

**Comment by tejlmand** [comment]
> but if using:
> 
> ```
> $ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild `samples/openthread/cli/
> $ ninja -Cbuild menuconfig
> # Set:
> # NRF_802154_SL_OPENSOURCE=y
> # NRF_802154_SOURCE_HAL_NORDIC=y
> ```

the problem seems to be that `CONFIG_MPSL` is still enabled when changing from `NRF_802154_SOURCE_NRFXLIB` to `NRF_802154_SOURCE_HAL_NORDIC` when menuconfig is used.

**Comment by e-rk** [comment]
@tejlmand This is a problem I was aware of previously. When selecting the the `CONFIG_NRF_802154_SL_OPENSOURCE=y` you must also explicitly set `CONFIG_MPSL=n`.

The MPSL/nrf-802154/BLE kconfigs are in a need of some refactoring to clean-up this use.

**Comment by tejlmand** [comment]
> @tejlmand This is a problem I was aware of previously. When selecting the the `CONFIG_NRF_802154_SL_OPENSOURCE=y` you must also explicitly set `CONFIG_MPSL=n`.
> 
> The MPSL/nrf-802154/BLE kconfigs are in a need of some refactoring to clean-up this use.

So you consider it unrelated to this PR ?

**Comment by e-rk** [comment]
It's been a problem that was present long before this PR and the scope affects MPSL and SD controller, which are beyond this PR.
It's something I would leave for the future.

**Comment by tejlmand** [comment]
> @tejlmand This is a problem I was aware of previously. When selecting the the `CONFIG_NRF_802154_SL_OPENSOURCE=y` you must also explicitly set `CONFIG_MPSL=n`.

I think this should be highlighted somewhere.

File: west.yml [hunk 5]
```diff
@@ -145,6 +147,12 @@ manifest:
       remote: nordicsemi
       revision: 24f1b2b0c64c694b7f9ac1b7eab60b39236ca0bf
       path: modules/lib/cddl-gen
+    - name: nrf-802154
```
[the remark this answers is not in the document]
Comment (e-rk) [reply west.yml]: Done.

**Comment by e-rk** [comment]
@tejlmand 

> I think this should be highlighted somewhere.

I have created a jira ticket to track this: https://projecttools.nordicsemi.no/jira/browse/KRKNWK-9681

**Review by hubertmis — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]
looks great

**Review by jciupis — APPROVED** [review]
LGTM :+1:

**Comment by carlescufi** [comment]
Merging as agreed, since the build failures are unrelated to this PR.

**Pull request merged**